src/java_bytecode/expr2java.h

```diff
@@ -50,4 +51,34 @@ class expr2javat:public expr2ct
 std::string expr2java(const exprt &expr, const namespacet &ns);
 std::string type2java(const typet &type, const namespacet &ns);
 
+/// Convert floating point number to a string without printing
+/// unnecessary zeros. Prints decimal if precision is not lost.
+/// Prints hexadecimal otherwise, and/or java-friendly NaN and Infinity
+template <typename float_type>
+std::string floating_point_to_java_string(float_type value)
+{
+  static const std::string class_name = std::is_same<float_type, float>::value
+    ? "Float"
+    : "Double";
+  if (value == std::numeric_limits<float_type>::quiet_NaN())
```

Comment:
Does `==` work for NaNs in C? I think you might need to use `std::isnan`. This would also remove the obligation on the caller to coerce NaNs to this particular bitpattern.